### PR TITLE
Wire submission questions into the quest submission flow (#1304 PR 2)

### PR DESCRIPTION
**Pull request by tylerecouture**
### What?
**PR 2 of the Additional and Multiple Submission Types feature (#1304)**, building on the `questions` app merged in #2085. For quests that have questions:

- **Students** see one answer field per question on the quest submission page (single-line short answer, Summernote long answer, or file upload, with per-question instructions and required/optional marking; the answer fields are label-less: the question's instructions, shown on the same line as the "Question N:" heading, act as the label, with aria-labels for accessibility). Text answers **autosave** with the draft; files upload at submit time. A general **"Additional Submission Comments"** box sits below the questions.
- **Completion is enforced:** required questions must be answered (with allowed-file-type checking for uploads) before the quest can be submitted; invalid answers re-render the page with each question's errors visible. Answered questions count as submission content, so students are no longer blocked by the "attach something or comment" rule when their answers *are* the submission.
- **Marking:** the published answers appear under the submission comment for both the student and the marker, in a table with aligned **Question / Your answer** headers (the marker sees **Answer provided**); staff additionally see each question's solution text/file and marker notes beside the student's answer.
- **Teachers** get a *Submission Questions* panel on every quest's detail page, captioned "(students answer these when submitting)", with a **Manage Questions** button and **detailed help text** explaining the feature: the three question types, required/optional semantics, autosave, the marking view, and what happens to answers when a question is deleted. The question list page carries the same expanded guidance.
- **The `enable_submission_questions` SiteConfig flag from #2085 is removed** (field + migration + view gating + tests). Research into the original #1304 plan confirmed no deck-level gate was ever intended: the natural per-quest gate is enough, so **a quest with no questions behaves exactly as it always has**, and the toggle only ever hid the staff CRUD pages. The guidance from its help text moved into the feature's own UI (expanded, per above).

### Why?
This is the integration layer that was deliberately cut from #2085: the original #1639 version of it conflicted with two years of `quest_manager` rewrites, so it's re-implemented here on top of the current FBVs (`submission`, `complete`, `ajax_save_draft`), carrying the fixes for the integration bugs found in the #1639 review. The flag removal restores the original design intent (see #1304's plan: questions are additive *"plus the current system"*, with question-less quests unchanged).

### How?
- **Draft-row reconciliation** (`questions/utils.py::sync_draft_question_submissions`): every time the answer formset is built (page render and complete POST), draft rows are synced against the quest's *current* questions. This fixes three #1639 review findings: questions added mid-draft now appear and are enforced; questions deleted mid-draft drop out without a 500 or a submission-blocking dead end; and a quest that gains its first question mid-draft still gets a working formset.
- **`complete()`**: builds the bound formset only for the "complete" action on a not-yet-completed submission (a duplicate POST via the back button can't spawn stray rows, and the "comment" quick-reply path is untouched). Validation happens before anything is written; on success the answers are saved and **published** by pointing them at the completion comment. After a teacher returns a quest, the next cycle gets fresh draft rows while the previous cycle's published answers stay attached to their comment: per-resubmission history, as designed in #2085.
- **`ajax_save_draft`**: accepts an `answers` JSON payload pairing each formset row's id with its response text; rows are ownership-checked (this user's submission, unpublished only, short/long answer questions only) and saved when changed. File answers deliberately don't autosave (#1639's file autosave silently discarded most files: wrong MIME sentinel handling, 2.5 MB in-memory cutoff), so files upload once, at submit, through the validated formset.
- **`submission.html`**: renders the formset (management form + per-question panels + visible non-form errors) above the comment field, and scopes the previously-global `div.note-editable` selectors to the main comment form: with multiple Summernote editors on the page, the old selectors would have autosaved a question answer as the comment draft and broken quick-text insertion.
- **`comments.html`**: a comment now renders its published answers (via the `question_submissions` reverse relation) through a new `questions/snippets/comment_answers.html`, so the display works on the submission page and the approvals previews alike, and announcement/other comments are unaffected (they have no answers).
- **Submission-question UI (review round):** short answer is a single-line input; the question number and instructions render on one line ("Question N: …") with the redundant "Instructions" label removed; a general "Additional Submission Comments" heading appears above the comment box when a quest has questions; the answer tables gain aligned Question / Your answer (marker: Answer provided) headers, **each answer is numbered (1., 2., …) before its prompt** to match the on-form "Question N" labels, and the **redundant "Question Answers" heading above the table was dropped** since the column headers already convey it; and marker notes render inline beside their label. A small `unwrap_p` template filter (`questions/templatetags/question_tags.py`, unit-tested) strips Summernote's wrapping `<p>` so a one-line prompt/note flows inline after its label instead of dropping onto its own row; multi-paragraph content is left untouched.
- **Answer-display polish:** each numbered answer shows a small Font Awesome type icon right after its number (text cursor = short answer, paragraph lines = long answer, paperclip = file upload; type name in a native `title` tooltip) via a unit-tested `question_type_icon` filter, so a marker can tell an answer's shape at a glance. And the answers table wraps long content (`table-layout: fixed` + `overflow-wrap`/`word-break`, `pre-wrap` for code) so a long URL or unbroken token can't widen its column and clip the other answers (see the layout stress test below).
- **Gating** is purely per-quest: `quest.question_set.exists()`. All views/templates/tests for the removed SiteConfig flag are gone (`siteconfig` migration 0032 drops the field).
- Develop is merged in, keeping its localized select2 CSS alongside the formset media, and playing nicely with the new comment-HTML sanitization and submission-page tooltips. (Latest merge also reconciles #2187's reorder UI, which landed on develop first: `QuestionMoveView` drops the now-removed feature-flag mixin.)
- One template quirk worth knowing for review: the formset loop variable must be named `form`, because `submission.html` references `{{ form.question.instructions }}` / `{{ form.question.required }}` directly to render each question's heading (commented in the template).

### Testing?
- Integration tests in `src/questions/tests/test_integration.py` covering: formset presence/absence (no questions, staff, awaiting approval), mid-draft question add/delete, completion success/enforcement/publish, resubmission cycles, autosave (save, ownership, published-row immutability, file-row text rejection, malformed/irrelevant payloads, no needless re-saves), the marking display (student vs staff visibility of solutions/notes, inline answer numbering, per-type icons, multi-paragraph marker notes), and the quest-detail entry point. Plus unit tests for the `unwrap_p` and `question_type_icon` filters.
- Full suite green locally, plus `ruff check src`, `manage.py check`, `makemigrations --check --dry-run` clean.
- The whole flow was exercised live in a real browser on a dev deck: answering, submitting, and marking.

### Screenshots
From a real `hackerspace` dev deck (student `angela.harris`, teacher `owner`).

**Student submission form** (single-line short answer, inline "Question N:", Summernote long answer, and the comment box with its new heading):
student form before/after

**Student's own submitted answers** (numbered, with a type icon after each number; Question / Your answer headers; no redundant "Question Answers" heading):
student answers

**Marker view** (numbered answers with type icons, Question / Answer provided headers, each question's Solution and Marker notes inline):
marker view

**Layout stress test** (marker view): a long, fully-formatted long answer (headings, bold/italic/underline/strike, colour, sub/superscript, lists, blockquote, code, a table and an image), the longest single-line short answer including an unbroken URL, and a real file attachment. This is what caught the wrap fix above: before, long/unbroken content widened the column and clipped answers at the panel edge; after, everything wraps within its cell.

_Before (long content overflows and is clipped):_
stress before

_After (wraps within the cell):_
stress after

Earlier feature-overview screenshots (quest-detail panels, no-questions parity, help text)

_Note: in this set the Summernote toolbar assets were CDN-blocked in the capture sandbox, so long-answer/comment fields show as bare textareas; they are Summernote editors in the app (see the answer views above)._

**Quest with NO questions - teacher:**
no questions teacher

**Quest with NO questions - student** (identical to today):
no questions student

**Question list page with the detailed help text:**
question list help text

**Quest detail page for staff, quest with questions** (panel + caption + Manage Questions):
quest detail help text

### Anything else?
With this PR the #1304 feature is usable end to end, with no configuration required: add questions to a quest, students answer while submitting, markers see answers beside solutions. Decks and quests without questions are unaffected. Not included (possible follow-ups): autosave for file answers, and importing questions with quests via the Shared Library. (Question reordering shipped separately in #2187.)

Part of #1304 (with #2085). Completes the salvage of #1639.

- Claude Code (`Bytedeck - multiple submission types`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_012Ptw8gWroqLnsLWieBaY67

## Summary by CodeRabbit

- **New Features**
  - Students can answer submission questions directly within quest forms.
  - Text answers autosave with drafts, while file answers submit with the quest.
  - Staff can manage questions from quest details and review answers alongside comments.
  - Answer displays are numbered, show a per-type icon, wrap long content, and include clearer instructions, solutions, and marker notes.

- **Bug Fixes**
  - Improved duplicate draft-answer handling and stale submission validation.
  - Added sanitization to prevent unsafe content in submitted answers.
  - Long answers and URLs no longer overflow the answers table.
  - Required questions are enforced, while optional unanswered questions support comment-based completion.

### CODE REVIEW

**Comment by tylerecouture** [comment]
@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/questions/views.py [hunk 1]
```diff
@@ -4,31 +4,14 @@
...
-
-    def dispatch(self, request, *args, **kwargs):
-        """404 the request when the feature flag is off."""
-        if not SiteConfig.get().enable_submission_questions:
-            raise Http404("Submission questions are not enabled on this deck.")
-        return super().dispatch(request, *args, **kwargs)
-
-
-class QuestionListView(NonPublicOnlyViewMixin, SubmissionQuestionsEnabledMixin, StaffMemberRequiredMixin, ListView):
+class QuestionListView(NonPublicOnlyViewMixin, StaffMemberRequiredMixin, ListView):
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/questions/views.py:16]: Same false positive the analyzer raised on #2085 (see https://github.com/bytedeck/bytedeck/pull/2085#discussion_r3628246815) — re-triggered here only because the flag-removal commit touched these four class-definition lines. Both mixins define `dispatch` cooperatively (`@method_decorator(...)` + `return super().dispatch(...)`), so the MRO chains them; the behavior is pinned by `test_all_question_page_status_codes__anonymous` / `__student` in src/questions/tests/test_views.py. No change needed; resolving the four identical threads.

- Claude Code (`Bytedeck - multiple submission types`)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by tylerecouture — COMMENTED** [review]
Some questions:

1. Provide screenshots of what the detail quest view looks like for teachers and students, when there are no submission questions.

2. Why doesn't it say "response text" for question 2? The long form answer?

3. Why does it say (visible to staff only) when the questions are visible to students too? Just the answers are staff only.

**Comment by tylerecouture** [comment]
All three addressed (fixes in `d2aae7c`, plus develop merged in via `e7e272f` to clear the conflict):

**1. Quest detail with no submission questions — live screenshots:**

*Teacher* — the panel appears with its help text, "No questions to display," and the Manage Questions button, so teachers can discover the feature from any quest:
no questions teacher

*Student* (their view of the same questionless quest, via their submission page which embeds the quest detail content) — no panel, no question fields, nothing new anywhere; identical to today:
no questions student

**2. "Response text" label inconsistency — fixed by removing it.** The short answer field had an explicit `label="Response text"` left over from PR 1, while the long answer (like the main comment box) renders label-less because the question's instructions sit directly above and act as the label. Rather than adding a redundant label to question 2, both are now label-less and consistent, with an `aria-label` on the short answer field for screen readers.

**3. "(visible to staff only)" — you're right, that was wrong.** The questions themselves are absolutely student-visible (they answer them when submitting); only this management panel and the solutions/marker notes are staff-only. The caption now reads **"(students answer these when submitting)"**, and the updated screenshots above/in the description reflect it.

- Claude Code (`Bytedeck - multiple submission types`)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/questions/templates/questions/snippets/comment_answers.html [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+          {% if request.user.is_staff %}
+            {% if answer.question.solution_text %}
+              <p class="text-muted"><small><b>Solution:</b> {{ answer.question.solution_text|linebreaksbr }}</small></p>
+            {% endif %}
+            {% if answer.question.solution_file %}
+              <p class="text-muted"><small><b>Solution file:</b>
+                <a href="{{ answer.question.solution_file.url }}" target="_blank">{{ answer.question.solution_file|filename }}</a></small></p>
+            {% endif %}
+            {% if answer.question.marker_notes %}
+              <p class="text-muted"><small><b>Marker notes:</b> {{ answer.question.marker_notes|safe }}</small></p>
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/questions/templates/questions/snippets/comment_answers.html]: Fixed in `0c01a2a` — good catch. `QuestionSubmissionForm.clean_response_text` now runs every answer through `sanitize_comment_html` (the allow-list sanitizer from #2113): short answers get stray/dangerous markup neutralized, and long answers keep their formatting but lose `on*` attributes that the "safe" summernote widget's tag filter alone lets through. Rendering stays `|safe` since values are sanitized at clean time; published values always pass through the form (the autosave path stores drafts raw, but drafts are only ever rendered back into the owner's own editor — same exposure model as existing comment drafts — and publishing re-cleans from the POST). Covered by two new tests: script tags neutralized in short answers, `onclick` stripped with `<b>` kept in long answers.

- Claude Code (`Bytedeck - multiple submission types`)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/questions/utils.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    existing_question_ids = set(
+        QuestionSubmission.objects.filter(
+            quest_submission=quest_submission, comment__isnull=True
+        ).values_list("question_id", flat=True)
+    )
+    QuestionSubmission.objects.bulk_create([
+        QuestionSubmission(quest_submission=quest_submission, question=question)
+        for question in questions
+        if question.id not in existing_question_ids
+    ])
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/questions/utils.py]: Addressed in `0c01a2a`, but with a different mechanism than suggested — a partial unique constraint turns out to be unsafe here: deleting a published comment reverts its answers to draft via `comment` SET_NULL, and if a new cycle's draft for the same question already exists, the SET_NULL itself would violate the constraint and turn a legitimate staff comment-deletion into an `IntegrityError` (adding the constraint made exactly that test fail). Instead, `sync_draft_question_submissions` now **self-heals**: on every render it collapses duplicate draft rows per question, keeping the contentful/most-recently-edited row and deleting the rest. That converges both the concurrent-first-render race and the comment-deletion revert path without a 500, with a test covering the healing. The reasoning is documented in the function's docstring.

- Claude Code (`Bytedeck - multiple submission types`)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by tylerecouture — COMMENTED** [review]
A few tweaks needed:

1. When there are Questions, then the submission form's general long answer has not context (not heading field), we need to add one to the top of it (e.g. Additional Submission Comments):
Image

Also, is there a reason the summernote wysiwyg widgets/toolbar doesn't load in the screenshots?

2. In the questions:

Image

a. The short answer should be a single line.  
b. Long answer should be a Summernote field.
c. Remove the label "respons text"
d. Put all the instructions on the same line to save vertical space, i.e: 
"Question 1: What is the url of your published website?"  (no need to put "Instructions")

3. In the student view of their own answers, the Question | Your Answer headings don't align with the columns (also change to "Your answer")

Image

4. In the teacher marking view, there is also misalignment of the headings (and use "Answer provided" instead of "Answer"

and save vertical space by putting the marker notes inline with the label "Marker notes: " to match the solution field.

Image

**Comment by tylerecouture** [comment]
Thanks — all addressed in `ac3c612`. Point by point:

**1. General comment box now has a heading.** When a quest has questions, an **"Additional Submission Comments"** heading now sits above the general comment editor, so its purpose is clear alongside the per-question fields.

> is there a reason the summernote wysiwyg widgets/toolbar doesn't load in the screenshots?

That was an artifact of my capture environment, not the app: the sandbox I screenshot in blocks the CDN font/asset the Summernote toolbar icons load from, so the editors rendered as bare textareas. In the real app they load fine — the new screenshots below are from a normal dev deck and show the full toolbar. (This also answers **2b**: the long-answer field was already a Summernote editor — it just looked like a textarea in the old screenshots.)

**2. Question fields:**
- **a.** Short answer is now a single-line text input (was a 2-row textarea).
- **b.** Long answer is a Summernote field (see above — already was; now visibly so).
- **c.** The "response text" label was already gone (set to empty back in `d2aae7c`); the old screenshots pre-dated that. Confirmed no label renders.
- **d.** The question number and instructions are now on one line — **"Question 1: What is the URL…?"** — and the redundant "Instructions" label is dropped.

**3. Student's own answers** — added aligned column headers **"Question" / "Your answer"**.

**4. Marker view** — added aligned **"Question" / "Answer provided"** headers, and the marker notes now render **inline beside their label** like the solution field.

To get 3 & 4 (and 2d) right I added a small `unwrap_p` template filter: Summernote wraps even a one-line note in a `<p>`, and nesting that block in an inline context makes the browser bump it onto its own line — the filter strips a single wrapping `<p>` so it flows inline (multi-paragraph content is left alone). Unit-tested.

All from a real `hackerspace` dev deck (student `angela.harris`, teacher `owner`):

**Student submission form** (single-line short answer, inline "Question N:", Summernote long answer + comment box with its new heading):
student form

**Student's submitted answers:**
student answers

**Marker view** (note the marker notes moving inline, and the aligned headers):
marker view

Full suite green (94 questions tests incl. the new filter tests + `SubmissionViewTests`), `ruff`/`check`/`makemigrations --check` clean.

- Claude Code (`Bytedeck - multiple submission types`)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/quest_manager/templates/quest_manager/submission.html [hunk 4]
```diff
@@ -79,7 +81,34 @@ <h3 class="panel-title">Quest Submission Form</h3>
...
+            <div class="well well-sm question-submission-form">
+              {# question number and the teacher's instructions on one line to save space; unwrap_p strips #}
+              {# summernote's wrapping <p> so a one-line prompt flows right after the "Question N:" label #}
+              <p>
+                <strong>Question {{ forloop.counter }}{% if not form.question.required %} <small>(optional)</small>{% endif %}:</strong>
+                {{ form.question.instructions|unwrap_p }}
+              </p>
+              {% crispy form %}
+            </div>
+          {% endfor %}
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/quest_manager/templates/quest_manager/submission.html:123]: Fixed at both sites in `9b0fd58`. Confirmed the diagnosis: `unwrap_p` only strips a *single* wrapping `<p>`, so multi-paragraph instructions/marker-notes keep their `<p>` tags — invalid nested in a `<p>`, and the browser auto-closes the outer one. Changed the wrappers to `<div>` (`submission.html` instructions, and `comment_answers.html` `.text-muted` marker-notes), keeping the inline "Question N: …" / "Marker notes: …" layout for the common single-paragraph case. Added a regression test (`AnswerDisplayTest.test_display__multiparagraph_marker_notes_render_in_div`) asserting the `<div class="text-muted">` wrapper with multi-paragraph notes.

- Claude Code (`Bytedeck - multiple submission types`)

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by tylerecouture** [comment]
Addressed the rest of the CodeRabbit review in `9b0fd58` (the outside-diff and nitpick items don't have inline threads to reply on):

- **`ajax_save_draft` autosave hardening** (Major, `views.py`) — the `answers` JSON is client-controlled. A valid-JSON non-object (e.g. `[]`) hit `.items()` → 500, and a non-integer `id` hit `filter(pk=…)` → 500. Now non-dict payloads coerce to `{}` and non-integer ids are skipped. Two new tests (`test_autosave__non_dict_answers_payload_ignored`, `test_autosave__non_integer_id_ignored`) fail without the guards.
- **Distinct `aria-label` per short-answer input** (nitpick, `forms.py`) — each short-answer input now gets `"Response to question N"` where **N matches the visible "Question N:" heading** (derived from the form's formset prefix, so it stays in sync rather than using the DB ordinal which can differ). Falls back to `"Response"` for a standalone form. Covered by `test_short_answer__distinct_aria_labels_in_formset` and `…__aria_label_fallback_without_formset`.
- **`QuestSubmission.get_comments` docstring** (nitpick, `models.py`) — added.

All new branches are at 100% coverage; full `questions` + `quest_manager` submission/complete suites, `ruff`, and `manage.py check` are green.

- Claude Code (`Bytedeck - multiple submission types`)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/questions/templates/questions/snippets/comment_answers.html [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+              <p class="text-muted"><small><b>Solution:</b> {{ answer.question.solution_text|linebreaksbr }}</small></p>
+            {% endif %}
+            {% if answer.question.solution_file %}
+              <p class="text-muted"><small><b>Solution file:</b>
+                <a href="{{ answer.question.solution_file.url }}" target="_blank">{{ answer.question.solution_file|filename|safe }}</a></small></p>
+            {% endif %}
+            {% if answer.question.marker_notes %}
+              {# unwrap_p strips summernote's wrapping <p> so a one-line note sits beside its label, like Solution above. #}
+              {# A <div> (not <p>) wraps it because multi-paragraph notes keep their <p> tags, invalid nested in a <p>. #}
+              <div class="text-muted"><small><b>Marker notes:</b> {{ answer.question.marker_notes|unwrap_p }}</small></div>
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/questions/templates/questions/snippets/comment_answers.html]: Declining this one, keeping the current markup. Two reasons:

1. **The HTMLHint error you cited was the literal `<p>` in the adjacent template *comment*, not this runtime markup** — that's fixed in 79980b4 (comments reworded to "paragraph element" / "div wrapper"), so the linter is now clean here.
2. The invalid nesting only bites **multi-paragraph** marker notes, which are rare — for the common single-paragraph note `unwrap_p` strips the wrapping `<p>`, so nothing block-level lands inside `<small>` and it's valid. Moving the note content outside `<small>` (as suggested) would render marker-notes text at **normal size**, out of step with the "Solution:" line right above it, which deliberately keeps its text in `<small>`. I'd rather keep those two visually consistent than change their size to satisfy a rare edge case browsers already render fine.

I did strengthen the multi-paragraph test (79980b4) to assert both `<p>` tags survive verbatim, so the behavior is pinned either way.

- Claude Code (`Bytedeck - multiple submission types`)

---
_Generated by [Claude Code](https://claude.ai/code)_
